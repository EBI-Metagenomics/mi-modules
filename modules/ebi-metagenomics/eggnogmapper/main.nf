process EGGNOGMAPPER {
    tag "$meta.id"
    label 'process_long'

    conda "bioconda::eggnog-mapper=2.1.12"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/eggnog-mapper:2.1.12--pyhdfd78af_0':
        'biocontainers/eggnog-mapper:2.1.12--pyhdfd78af_0' }"

    input:
    tuple val(meta), path(fasta)
    tuple val(meta2), path(annotation_hit_table)
    path(eggnog_data_dir)
    path(eggnog_db)
    path(eggnog_diamond_db)

    output:
    tuple val(meta), path("*.hits")   , emit: hits, optional: true
    tuple val(meta), path("*.annotations")    , emit: annotations, optional: true
    tuple val(meta2), path("*.seed_orthologs"), emit: orthologs, optional: true
    path "versions.yml"                       , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    def annotation_hit_input = annotation_hit_table ? "--annotate_hits_table ${annotation_hit_table}" : ""
    def eggnog_db_input = eggnog_db ? "--database ${eggnog_db}" : ""
    def eggnog_diamond_db_input = eggnog_diamond_db ? "--dmnd_db ${eggnog_diamond_db}" : ""
    def dbmem = task.memory.toMega() > 40000 ? '--dbmem' : ''

    def fasta_bool = fasta ? fasta.name : "no_fasta"
    def is_compressed = fasta_bool.endsWith(".gz")
    def fasta_name = fasta_bool.replace(".gz", "")
    def fasta_input = fasta ? "-i ${fasta_name}" : ""

    """
    if [ "$is_compressed" == "true" ]; then
        gzip -c -d $fasta > $fasta_name
    fi

    emapper.py \\
        ${args} \\
        ${fasta_input} \\
        ${annotation_hit_input} \\
        ${eggnog_db_input} \\
        ${eggnog_diamond_db_input} \\
        ${dbmem} \\
        --cpu ${task.cpus} \\
        --data_dir ${eggnog_data_dir} \\
        --output ${prefix}

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        eggnog-mapper: \$(echo \$(emapper.py --version) | grep -o "emapper-[0-9]\\+\\.[0-9]\\+\\.[0-9]\\+" | sed "s/emapper-//")
    END_VERSIONS
    """

    stub:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    touch ${prefix}.emapper.hits

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        eggnog-mapper: \$(echo \$(emapper.py --version) | grep -o "emapper-[0-9]\\+\\.[0-9]\\+\\.[0-9]\\+" | sed "s/emapper-//")
    END_VERSIONS
    """
}
