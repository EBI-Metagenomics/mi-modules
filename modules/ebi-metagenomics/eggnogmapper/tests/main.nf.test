
nextflow_process {

    name "Test Process EGGNOGMAPPER"
    script "../main.nf"
    process "EGGNOGMAPPER"

    tag "modules"
    tag "modules_nfcore"
    tag "diamond/makedb"
    tag "eggnogmapper"

    test("test_eggnogmapper_withdiamond") {
        setup {
            run("DIAMOND_MAKEDB"){
                script "../../diamond/makedb/main.nf"
                process {
                    """
                    input[0] = [
                        [ id:'test', single_end:false ], // meta map
                        file("${moduleDir}/tests/fixtures/test.fasta", checkIfExists: true)
                        ]
                    """
                }
            }
        }
        when {
            process {
                config "./withdiamond.nextflow.config"

                """                
                input[0] = [
                    [ id:'test', single_end:false ], // meta map
                    file("${moduleDir}/tests/fixtures/test.fasta", checkIfExists: true)
                    ]
                input[1] = [
                    [ id:'test', single_end:false ], // meta map
                    []
                    ]
                input[2] = file("${moduleDir}/tests/fixtures", type: 'dir', checkIfExists: true)
                input[3] = file("${moduleDir}/tests/fixtures/eggnog.db", checkIfExists: true)
                input[4] = DIAMOND_MAKEDB.out.db.map{ it[1] }
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

    test("test_eggnogmapper_withoutdiamond") {
        setup {
            run("DIAMOND_MAKEDB"){
                script "../../diamond/makedb/main.nf"
                process {
                    """
                    input[0] = [
                        [ id:'test', single_end:false ], // meta map
                        file("${moduleDir}/tests/fixtures/test.fasta", checkIfExists: true)
                        ]
                    """
                }
            }
        }
        when {
            process {
                config "./withoutdiamond.nextflow.config"

                """                
                input[0] = [
                    [ id:'test', single_end:false ], // meta map
                    []
                    ]
                input[1] = [
                    [ id:'test', single_end:false ], // meta map
                    file("${moduleDir}/tests/fixtures/test.hits", checkIfExists: true)
                    ]
                input[2] = file("${moduleDir}/tests/fixtures", type: 'dir', checkIfExists: true)
                input[3] = []
                input[4] = []
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

}
