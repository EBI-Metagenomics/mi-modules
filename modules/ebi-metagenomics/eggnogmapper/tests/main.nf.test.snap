{
    "test_eggnogmapper_withoutdiamond": {
        "content": [
            {
                "0": [
                    
                ],
                "1": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.emapper.annotations:md5,acf16652adebcb33083edd74c5ec6e5f"
                    ]
                ],
                "2": [
                    
                ],
                "3": [
                    "versions.yml:md5,d3e4efad28b5a924585ea3dfcf72c32c"
                ],
                "annotations": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.emapper.annotations:md5,acf16652adebcb33083edd74c5ec6e5f"
                    ]
                ],
                "csv": [
                    
                ],
                "orthologs": [
                    
                ],
                "versions": [
                    "versions.yml:md5,d3e4efad28b5a924585ea3dfcf72c32c"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.04.3"
        },
        "timestamp": "2024-07-03T14:39:05.516376238"
    },
    "test_eggnogmapper_withdiamond": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.emapper.hits:md5,86e4f5944b48e35533e14bb0518064e6"
                    ]
                ],
                "1": [
                    
                ],
                "2": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.emapper.seed_orthologs:md5,4535c05b5da52ec45c02d1df0176fc42"
                    ]
                ],
                "3": [
                    "versions.yml:md5,d3e4efad28b5a924585ea3dfcf72c32c"
                ],
                "annotations": [
                    
                ],
                "csv": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.emapper.hits:md5,86e4f5944b48e35533e14bb0518064e6"
                    ]
                ],
                "orthologs": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.emapper.seed_orthologs:md5,4535c05b5da52ec45c02d1df0176fc42"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,d3e4efad28b5a924585ea3dfcf72c32c"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.04.3"
        },
        "timestamp": "2024-07-03T14:15:10.109814679"
    }
}